x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -862,7 +874,18 @@ public Collection<Object> createComponents(
 
         this.mlUpgradeModeActionFilter.set(new MlUpgradeModeActionFilter(clusterService));
 
-        MlIndexTemplateRegistry registry = new MlIndexTemplateRegistry(settings, clusterService, threadPool, client, xContentRegistry);
+        if (machineLearningExtension.get() == null) {
+            machineLearningExtension.set(new DefaultMachineLearningExtension());
+        }
```

Comment:
I don't think this should be done here. You are assuming that `createComponents` gets called before any other method on the class that might need to behave differently depending on extensions. At the moment that's true, but as we add more extension points it might not be.

See my previous comment for what I think should be done instead.